app/src/main/kotlin/com/wire/android/util/logging/LogFileWriterV2Impl.kt

```diff
@@ -0,0 +1,374 @@
+                    val shouldFlush = logBuffer.size >= config.maxBufferSize ||
+                        ((currentTime - lastFlushTime) >= config.flushIntervalMs)
+
+                    if (shouldFlush) {
+                        flushBuffer()
+                        lastFlushTime = currentTime
+                    }
+
+                    return@withLock activeLoggingFile.length()
+                }
+            }
+        } catch (e: TimeoutCancellationException) {
+            appLogger.w("Buffer write operation timed out, log line may be lost: $text")
+            // Return current file length as fallback
+            return@withContext activeLoggingFile.length()
+        } catch (e: Exception) {
+            appLogger.e("Error writing to log buffer", e)
+            return@withContext activeLoggingFile.length()
+        }
+    }
+
+    private fun ensureLogDirectoryAndFileExistence() {
+        if (!logsDirectory.exists() && !logsDirectory.mkdirs()) {
+            appLogger.e("Unable to create logs directory")
+        }
+
+        if (!activeLoggingFile.exists() && !activeLoggingFile.createNewFile()) {
+            appLogger.e("KaliumFileWriter: Failure to create new file for logging", IOException("Unable to load log file"))
+        }
+        if (!activeLoggingFile.canWrite()) {
+            appLogger.e("KaliumFileWriter: Logging file is not writable", IOException("Log file not writable"))
+        }
+    }
+
+    override fun deleteAllLogFiles() {
+        clearActiveLoggingFileContent()
+        logsDirectory.listFiles()?.filter {
+            it.extension.lowercase(Locale.ROOT) == LOG_COMPRESSED_FILE_EXTENSION
+        }?.forEach { it.delete() }
+    }
+
+    private fun getCompressedFilesList() = (logsDirectory.listFiles() ?: emptyArray()).filter { it != activeLoggingFile }
+
+    private fun compressedFileName(): String {
+        val currentDate = logFileTimeFormat.format(Date())
+        return "${LOG_FILE_PREFIX}_$currentDate.$LOG_COMPRESSED_FILE_EXTENSION"
+    }
+
+    private fun deleteOldCompressedFiles() = getCompressedFilesList()
+        .sortedBy { it.lastModified() }
+        .dropLast(LOG_COMPRESSED_FILES_MAX_COUNT)
+        .forEach {
+            it.delete()
+        }
+
+    private suspend fun compressAsync() = withContext(Dispatchers.IO) {
+        try {
+            val compressedFile = File(logsDirectory, compressedFileName())
+
+            GZIPOutputStream(compressedFile.outputStream().buffered()).use { gzipOut ->
+                activeLoggingFile.inputStream().buffered().use { input ->
+                    input.copyTo(gzipOut, config.bufferSizeBytes)
+                }
+            }
+
+            appLogger.i("Log file compressed: ${activeLoggingFile.name} -> ${compressedFile.name}")
+        } catch (e: Exception) {
+            appLogger.e("Failed to compress log file: ${activeLoggingFile.name}", e)
+        }
+    }
+
+    private fun flushBuffer() {
```

Comment:
I'm wondering here, in general, if this function can be improved to avoid creating the secondary list `toList()` in place.


```suggestion
    private fun flushBuffer() {
        if (logBuffer.isEmpty()) return

        try {
            // Use BufferedWriter for efficient writing
            val writer = bufferedWriter ?: BufferedWriter(
                FileWriter(activeLoggingFile, true),
                config.bufferSizeBytes
            ).also { bufferedWriter = it }

            // Like here one, Write directly from buffer without copying to the .toList()
            logBuffer.forEach { line ->
                writer.appendLine(line)
            }
            writer.flush()

            // and here two, Clear only after successful write
            logBuffer.clear()

        } catch (e: IOException) {
            appLogger.e("Failed to flush log buffer", e)
        }
    }
```